Add an option to get an example pipeline

DimedS opened the pull request:
> **NOTE:** Kedro datasets are moving from `kedro.extras.datasets` to a separate `kedro-datasets` package in
> [`kedro-plugins` repository](https://github.com/kedro-org/kedro-plugins). Any changes to the dataset implementations
> should be done by opening a pull request in that repository.
## Description
This PR introduces an `--example` flag to the `kedro new` command. Users can specify `--example y/n` directly in the command line to indicate their preference. If not specified, users will be prompted interactively, with 'No' as the default selection.

User flow:
1. When a user selects Viz, PySpark, or both, the corresponding starter will be provided.
     If `example=No` is chosen in this scenario, data and config will be removed via hooks in the selected starter.
2. If neither Viz nor PySpark is selected:
   If `example=Yes` is chosen, the 'spaceflights-starter' will be provided.
   If `example=No` is chosen, a standard template will be provided.
3. Then final template will be stripped according to add_ons 1-5 selection, exception: Data folder, it will not be stripped if `example=Yes` chosen.

## Development notes

1. Temporary added a checkout for "3076-add-example-pipeline-to-hooks" branch of starters git repo to properly test together with changes in starters in PR [181](https://github.com/kedro-org/kedro-starters/pull/181) (643-663 lines in starters.py)
2. Will add additional unit tests to cover my changes.


## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

DimedS commented:
Do we need to copy postGen cookiecutter hooks to 4 spaceflights starters to enable `stripping` based on add_ons chosen, or we already have a better solution for that?
@merelcht  @noklam @ankatiyar @SajidAlamQB

merelcht commented on kedro/framework/cli/starters.py:
```diff
@@ -586,8 +609,9 @@ def _make_cookiecutter_args(
+    example_pipeline = extra_context.get("example_pipeline")
     starter_path = "git+https://github.com/kedro-org/kedro-starters.git"
-    if add_ons:
+    if example_pipeline:
```
This isn't correct. Even if the example pipeline isn't wanted, we might need to fetch from a starter template other than the default. @SajidAlamQB can probably help out here about what the flow should be.

DimedS replied:
@merelcht , thank you. You right, I need to consult with @SajidAlamQB what should we do in case of Spark and Viz and no example option.

SajidAlamQB replied:
I think this doesn't need to be changed as we already select the right template with examples in them for pyspark, viz and pyspark & viz. These templates already contain the example pipelines but are removed in `hooks/utils.py` `_handle_starter_setup`. So the current default case is `example=no`. I think we just need a check in `utils.py` to skip `_handle_starter_setup` if `example=yes`.

SajidAlamQB replied:
I think for the default template [1-5], options without pyspark and viz, we don't already have example pipelines so we might need to change template to `spaceflights-pandas` in here if `example=yes`.

DimedS replied:
> I think only for the default template [1-5], options without pyspark and viz, we don't already have example pipelines so we might need to change template to `spaceflights-pandas` in here if `example=yes`.

yes, looks like logic should be different for pyspark, viz and without them:
1. without pyspark, viz: If --example==yes, use sfaceflights-pandas starter, elif --example ==no, use default template
2. with pyspark or viz or both: use starter anyway, but we need modifications in  _handle_starter_setup

Is it correct?

SajidAlamQB replied:
Yes, this sounds right.

noklam commented:
https://github.com/kedro-org/kedro/issues/3054

@DimedS 
I notice the current order doesn't match with #3054, the example prompt should happen before `project_name`
![image](https://github.com/kedro-org/kedro/assets/18221871/f2b755d1-ce7a-42c0-b998-e887197c7359)

In addition, I wonder is `--example y/n` needed? Can it be reduced to `--example` since it is a boolean option or is it unclear? @amandakys

DimedS commented:
> #3054
> 
> @DimedS I notice the current order doesn't match with #3054, the example prompt should happen before `project_name` 
> 
> In addition, I wonder is `--example y/n` needed? Can it be reduced to `--example` since it is a boolean option or is it unclear? @amandakys

Hm, I see that the last version of #3054 with --example on the end:
https://github.com/kedro-org/kedro/issues/3054#issuecomment-1775659975

noklam commented:
I see - I didn't saw the latest comment. We should push the project name as the first prompt, but it's out of scope for this PR.

noklam commented on MANIFEST.in:
```diff
@@ -3,4 +3,4 @@ include LICENSE.md
 include kedro/ipython/*.png
 include kedro/ipython/*.svg
-recursive-include templates *
+recursive-include kedro/templates *
```
is this intended?

noklam reviewed (commented):
First of all, thank you for the PR and the detailed description. Having it checkout to a feature starter branch save me a lot of time to review and test.

I leave some comments but I think the key questions are
1. Can we simplify `example_pipeline` as a boolean flag from the CLI with just `--example`? If not, we should parse this early on as a boolean save ourselves from checking `example_pipeline` is a None or not at every step, since I think they don't make a difference.
2. Validation - the validation function assume `click` and it also does parsing, I think should keep it separated.

DimedS commented:
> 1. Can we simplify `example_pipeline` as a boolean flag from the CLI with just `--example`? If not, we should parse this early on as a boolean save ourselves from checking `example_pipeline` is a None or not at every step, since I think they don't make a difference.
> 2. Validation - the validation function assume `click` and it also does parsing, I think should keep it separated.
@noklam , many thanks for your notes.
I agree with all you written: 
1. If we can use just `--example` flag not `--example='y/n'` in CLI, it will be simple - less validation and parsing, but I think we cannot do that - if client will not provide `--example` flag in CLI it doesn't mean that he don't want example (we cannot put False here) - we need to ask him in Prompt. So I think we should leave it the same to provide a possibility to write 'Not' in CLI.
2. I'm also unhappy with current validation situation, I tried to put CLI validation directly to @click.option in click.Choice(["y", "n", "yes", "no"] and parsing in callback=_validate_yn func (seems to be a wrong function naming better name that func _parsing, but validation and parsing are separated currently). But it seems unbeautiful because of specific callback parsing function format that you mentioned.
And I agree that it's better to unify validation process (also wrote about it [there](https://github.com/kedro-org/kedro/issues/3264#issuecomment-1808423550)) because it's 3 possible inputs (CLI, config, prompts) better to validate and parse them in the same code if it's possible.
And you right, currently I don't provide a validation for loading from config, only parsing in _validate_yn().
I will rewrite validation and parsing part.

SajidAlamQB commented on kedro/templates/project/hooks/utils.py:
```diff
@@ -144,7 +144,7 @@ def _handle_starter_setup(selected_add_ons_list: str, python_package_name: str)
+def setup_template_add_ons(selected_add_ons_list: str, requirements_file_path: str, pyproject_file_path: str, python_package_name: str, example_pipeline: str) -> None:
     """Setup the templates according to the choice of add-ons.
 
     Args:
```
Don't forget to update docstrings with new arguments, `example_pipeline`.

merelcht commented on kedro/framework/cli/starters.py:
```diff
@@ -482,7 +504,10 @@ def _convert_addon_names_to_numbers(selected_add_ons_flag: str | None) -> str |
+    prompts_required: dict,
+    selected_add_ons_flag: str,
+    project_name: str,
+    example_pipeline: bool,
```
To me `example_pipeline` is fine for a variable name that is a boolean.

merelcht commented on kedro/framework/cli/starters.py:
```diff
@@ -495,6 +520,8 @@ def _select_prompts_to_display(
             or None in case the flag wasn't used.
         project_name: a string containing the value for the --name flag, or
             None in case the flag wasn't used.
+        example_pipeline: True or False for --example flag, or
+            None in case the flag wasn't used.
```
I think it needs to be `None` when `--example` wasn't used, because that means the user needs to be prompted interactively for the example choice.

merelcht commented on kedro/templates/project/cookiecutter.json:
```diff
@@ -3,5 +3,6 @@
     "python_package": "{{ cookiecutter.project_name.strip().replace(' ', '_').replace('-', '_').lower() }}",
     "kedro_version": "{{ cookiecutter.kedro_version }}",
-    "add_ons": "none"
+    "add_ons": "none",
+    "example_pipeline": "No"
```
It looks inconsistent to me that the default for `add_ons` is `none` with a lower case and for `example_pipeline` it's `No` with a capital letter. I'd just use `no` without capital letter.

merelcht commented on kedro/templates/project/hooks/utils.py:
```diff
@@ -169,13 +169,13 @@ def setup_template_add_ons(selected_add_ons_list: str, requirements_file_path: s
-    if "Data Structure" not in selected_add_ons_list:
+    if "Data Structure" not in selected_add_ons_list and example_pipeline != "True":
         _remove_dir(current_dir / "data")
 
-    if "Pyspark" in selected_add_ons_list:
+    if "Pyspark" in selected_add_ons_list and example_pipeline != "True":
         _handle_starter_setup(selected_add_ons_list, python_package_name)
 
-    if "Kedro Viz" in selected_add_ons_list:
+    if "Kedro Viz" in selected_add_ons_list and example_pipeline != "True":
```
I totally agree with @noklam here. I would rename `_handle_starter_setup` to make it clear this is removal of files, and I also think the code inside `_handle_starter_setup` can be more generic. We don't need to check pipeline names and test names, but just remove all files inside the pipelines folder etc.

DimedS replied:
I updated that part to clarify that it's only about Viz and PySpark, @merelcht do you think we should remove all files in tests/pipelines except of __init__.py?

merelcht reviewed (commented):
This part of the code https://github.com/kedro-org/kedro/blob/8c6d5c105df3f0982fec84ad74fdf23037664dad/kedro/framework/cli/starters.py#L862-L868 should also be updated. Now you don't see which add-ons you have selected when you use an example.

DimedS commented on kedro/framework/cli/starters.py:
```diff
@@ -168,13 +173,22 @@ def starter():
+    "-e",
+    "example_pipeline",
+    type=click.Choice(["y", "n", "yes", "no"], case_sensitive=False),
+    callback=_validate_yn,
```
Updated and centralized the validation process. If this looks effective, we can apply a similar approach to other regex validations.
Left --example='y/n' option unchanged

DimedS commented:
It appears that I've addressed all the review comments and have also included new tests to cover the `example` option. Could you please review it again? Currently, I'm noticing an issue with the docs, I will investigate it further, other tests are ok.

merelcht commented on kedro/framework/cli/starters.py:
```diff
@@ -143,6 +144,27 @@ class KedroStarterSpec:  # noqa: too-few-public-methods
+}
+
+
+def _validate_regex(pattern_name, text):
```
`_validate_regex` is maybe a bit of a misleading name, because it sounds like you can use this generically for regex patterns, but it's actually quite specific for the yes/no choice. I'd rename it to something like `_validate_yes_no` or `_validate_example_choice`

DimedS replied:
hm, the idea was that it's a general func to validate any regex pattern, and we should use it also for other regex validations, that currently hardcoded and duplicated in starters.py

merelcht replied:
Ahh I see now. So the idea is to also expand `VALIDATION_PATTERNS` to contain other patterns than yes/no? In that case ignore my comment :)

merelcht commented on kedro/templates/project/hooks/utils.py:
```diff
@@ -105,11 +105,11 @@ def _remove_file(path: Path) -> None:
 
 
-def _handle_starter_setup(selected_add_ons_list: str, python_package_name: str) -> None:
+def _remove_PySpark_Viz_starter_files(is_Viz: bool, python_package_name: str) -> None:
```
Nit: it looks a bit weird to have the capital letters in the function name. I'd personally just keep it lower case. Same for the `is_Viz` argument

merelcht commented on tests/framework/cli/test_starters.py:
```diff
@@ -82,10 +89,27 @@ def _get_expected_files(add_ons: str):
 
     for add_on in add_ons_list:
         expected_files = expected_files + add_ons_template_files[add_on]
+    # If example pipeline was choosen we don't need to delete /data folder
```
```suggestion
    # If example pipeline was chosen we don't need to delete /data folder
```

merelcht reviewed (commented):
> This part of the code
> 
> https://github.com/kedro-org/kedro/blob/8c6d5c105df3f0982fec84ad74fdf23037664dad/kedro/framework/cli/starters.py#L862-L868
> 
> should also be updated. Now you don't see which add-ons you have selected when you use an example.

@DimedS This still needs to be updated. If I choose an example I don't get a message showing me what `add_ons` I chose.

merelcht reviewed (approved):
This looks good to be merged now! Nice work @DimedS 👏 
Don't forget to add it to the release notes 😃

SajidAlamQB commented on tests/framework/cli/test_starters.py:
```diff
@@ -82,10 +89,27 @@ def _get_expected_files(add_ons: str):
+        expected_files += sum(example_files_count)
+        expected_files += (
+            4 if "7" in add_ons_list else 0
+        )  # add 3 .py and 1 parameters files in reporting for Viz
```
Thanks for the comments these tests can be confusing to understand. 😁

SajidAlamQB reviewed (approved):
Tested this manually and all looks good! Thanks you @DimedS awesome work! 🌟

noklam reviewed (approved):
Thanks for the PR! This is one of the more complicated piece.

Don't remember to update the RELEASE.md and revert the hardcoded branch.

```
        "checkout"
        ] = "3076-add-example-pipeline-to-hooks" 
```
